mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -122,6 +122,7 @@ private[spark] object RandomForest extends Logging with Serializable {
     timer.start("init")
 
     val retaggedInput = input.retag(classOf[Instance])
+    retaggedInput.persist()
```

Comment:
Here `retaggedInput` is not required to persist since only action applied on it, is in `buildMetadata()` and even `baggedInput` is already persisted.

Reply:
In findSplits(), there is another action collectAsMap applied on retaggedInput.

Reply:
> We add these internal persist calls for important partial results though, not user inputs, in general, and only if the user input was cached.

`retaggedInput` is just transformed user data (`input`) and further its partial result (`baggedInput`) is persisted at line 158.